failing to install with msys2/mingw nodejs

```
# npm i @napi-ffi/ffi-napi
npm warn cleanup Failed to remove some directories [
npm warn cleanup   [
npm warn cleanup     '\\\\?\\E:\\msys64\\home\\topkek\\git\\test\\node_modules\\@napi-ffi\\ref-napi',
npm warn cleanup     [Error: EBUSY: resource busy or locked, rmdir 'E:\msys64\home\topkek\git\test\node_modules\@napi-ffi\ref-napi'] {
npm warn cleanup       errno: -4082,
npm warn cleanup       code: 'EBUSY',
npm warn cleanup       syscall: 'rmdir',
npm warn cleanup       path: 'E:\\msys64\\home\\topkek\\git\\test\\node_modules\\@napi-ffi\\ref-napi'
npm warn cleanup     }
npm warn cleanup   ],
npm warn cleanup   [
npm warn cleanup     '\\\\?\\E:\\msys64\\home\\topkek\\git\\test\\node_modules\\@napi-ffi',
npm warn cleanup     [Error: EBUSY: resource busy or locked, rmdir 'E:\msys64\home\topkek\git\test\node_modules\@napi-ffi\ref-napi'] {
npm warn cleanup       errno: -4082,
npm warn cleanup       code: 'EBUSY',
npm warn cleanup       syscall: 'rmdir',
npm warn cleanup       path: 'E:\\msys64\\home\\topkek\\git\\test\\node_modules\\@napi-ffi\\ref-napi'
npm warn cleanup     }
npm warn cleanup   ],
npm warn cleanup   [
npm warn cleanup     '\\\\?\\E:\\msys64\\home\\topkek\\git\\test\\node_modules',
npm warn cleanup     [Error: EBUSY: resource busy or locked, rmdir 'E:\msys64\home\topkek\git\test\node_modules\@napi-ffi\ref-napi'] {
npm warn cleanup       errno: -4082,
npm warn cleanup       code: 'EBUSY',
npm warn cleanup       syscall: 'rmdir',
npm warn cleanup       path: 'E:\\msys64\\home\\topkek\\git\\test\\node_modules\\@napi-ffi\\ref-napi'
npm warn cleanup     }
npm warn cleanup   ]
npm warn cleanup ]
npm error code 1
npm error path E:\msys64\home\topkek\git\test\node_modules\@napi-ffi\ffi-napi
npm error command failed
npm error command C:\WINDOWS\system32\cmd.exe /d /s /c node-gyp-build
npm error make: Entering directory '/home/topkek/git/test/node_modules/@napi-ffi/ffi-napi/build'
npm error   CC(target) Release/obj.target/nothing/../../node-addon-api/nothing.o
npm error   AR(target) Release/obj.target/../../node-addon-api/nothing.a
npm error   COPY Release/nothing.a
npm error   RULE deps_libffi_libffi_gyp_ffi_target_preprocess_asm_0 Release/obj.target/ffi/geni/win64_intel.asm
npm error make: Leaving directory '/home/topkek/git/test/node_modules/@napi-ffi/ffi-napi/build'
npm error (node:2820) [DEP0190] DeprecationWarning: Passing args to a child process with shell option true can lead to security vulnerabilities, as the arguments are not escaped, only concatenated.
npm error (Use `node --trace-deprecation ...` to show where the warning was created)
npm error gyp info it worked if it ends with ok
npm error gyp info using node-gyp@11.2.0
npm error gyp info using node@24.8.0 | win32 | x64
npm error gyp info find Python using Python version 3.12.12 found at "E:/msys64/ucrt64/bin/python3.exe"
npm error gyp http GET https://nodejs.org/download/release/v24.8.0/node-v24.8.0-headers.tar.gz
npm error gyp http 200 https://nodejs.org/download/release/v24.8.0/node-v24.8.0-headers.tar.gz
npm error gyp http GET https://nodejs.org/download/release/v24.8.0/SHASUMS256.txt
npm error gyp http 200 https://nodejs.org/download/release/v24.8.0/SHASUMS256.txt
npm error gyp info spawn E:/msys64/ucrt64/bin/python3.exe
npm error gyp info spawn args [
npm error gyp info spawn args 'E:/msys64/ucrt64/lib/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py',
npm error gyp info spawn args 'binding.gyp',
npm error gyp info spawn args '-f',
npm error gyp info spawn args 'make',
npm error gyp info spawn args '-I',
npm error gyp info spawn args 'E:/msys64/home/topkek/git/test/node_modules/@napi-ffi/ffi-napi/build/config.gypi',
npm error gyp info spawn args '-I',
npm error gyp info spawn args 'E:/msys64/ucrt64/lib/node_modules/npm/node_modules/node-gyp/addon.gypi',
npm error gyp info spawn args '-I',
npm error gyp info spawn args 'E:/msys64/ucrt64/include/node/common.gypi',
npm error gyp info spawn args '-Dlibrary=shared_library',
npm error gyp info spawn args '-Dvisibility=default',
npm error gyp info spawn args '-Dnode_root_dir=E:/msys64/ucrt64',
npm error gyp info spawn args '-Dnode_gyp_dir=E:/msys64/ucrt64/lib/node_modules/npm/node_modules/node-gyp',
npm error gyp info spawn args '-Dnode_lib_file=E:/msys64/ucrt64/lib/libnode.dll.a',
npm error gyp info spawn args '-Dmodule_root_dir=E:/msys64/home/topkek/git/test/node_modules/@napi-ffi/ffi-napi',
npm error gyp info spawn args '-Dnode_engine=v8',
npm error gyp info spawn args '--depth=.',
npm error gyp info spawn args '--no-parallel',
npm error gyp info spawn args '--generator-output',
npm error gyp info spawn args 'build',
npm error gyp info spawn args '-Goutput_dir=.'
npm error gyp info spawn args ]
npm error gyp info spawn make
npm error gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
npm error /bin/sh: line 1: ../../../deps/libffi/preprocess_asm.cmd: No such file or directory
npm error make: *** [deps/libffi/ffi.target.mk:10: Release/obj.target/ffi/geni/win64_intel.asm] Error 127
npm error gyp ERR! build error
npm error gyp ERR! stack Error: `make` failed with exit code: 2
npm error gyp ERR! stack at ChildProcess.<anonymous> (E:\msys64\ucrt64\lib\node_modules\npm\node_modules\node-gyp\lib\build.js:219:23)
npm error gyp ERR! stack at ChildProcess.emit (node:events:508:28)
npm error gyp ERR! stack at ChildProcess._handle.onexit (node:internal/child_process:294:12)
npm error gyp ERR! System MINGW32_NT-10.0 10.0.26100
npm error gyp ERR! command "E:\\msys64\\ucrt64\\bin\\node.exe" "E:\\msys64\\ucrt64\\lib\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
npm error gyp ERR! cwd E:\msys64\home\topkek\git\test\node_modules\@napi-ffi\ffi-napi
npm error gyp ERR! node -v v24.8.0
npm error gyp ERR! node-gyp -v v11.2.0
npm error gyp ERR! not ok
npm error A complete log of this run can be found in: C:\Users\topkek\AppData\Local\npm-cache\_logs\2025-10-26T19_54_52_214Z-debug-0.log
```


although this is fixed in my fork https://github.com/node-ffi-napi/node-ffi-napi/compare/master...Kreijstal:node-ffi-napi:master look at the deps/libffi/libffi.gyp diffs

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -60,7 +60,7 @@ jobs:
             mingw-w64-x86_64-nodejs
             mingw-w64-x86_64-python
             mingw-w64-x86_64-gcc
-            mingw-w64-x86_64-make
+            make
 
       - name: Verify MinGW Node.js is active
         shell: msys2 {0}

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -42,3 +42,38 @@ jobs:
       - name: Prebuild
         shell: bash
         run: npm run prebuild
+
+  test-mingw:
+    name: Test MinGW (MSYS2 MINGW64) on Windows
+    runs-on: windows-latest
+
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+
+      - name: Set up MSYS2 with MINGW64
+        uses: msys2/setup-msys2@v2
+        with:
+          msystem: MINGW64
+          update: true
+          install: >-
+            mingw-w64-x86_64-nodejs
+            mingw-w64-x86_64-python
+            mingw-w64-x86_64-gcc
+            mingw-w64-x86_64-make
+
+      - name: Verify MinGW Node.js is active
+        shell: msys2 {0}
+        run: node -e "const os = require('os'); console.log('os.type():', os.type()); if (!os.type().startsWith('MINGW')) process.exit(1)"
+
+      - name: Install dependencies
+        shell: msys2 {0}
+        run: npm install
+
+      - name: Build native tests
+        shell: msys2 {0}
+        run: npm run build
+
+      - name: Run tests
+        shell: msys2 {0}
+        run: npm test

--- a/deps/libffi/libffi.gyp
+++ b/deps/libffi/libffi.gyp
@@ -9,10 +9,13 @@
 {
   'variables': {
     'target_arch%': 'ia32', # built for a 32-bit CPU by default
-    # Distinguish MSVC from MinGW/MSYS2 on Windows. MinGW reports an OS type
-    # starting with "MINGW" and uses GCC toolchain, so .S files compile directly
+    # Distinguish MSVC from MinGW/MSYS2 on Windows. MinGW-compiled Node reports
+    # an OS type starting with "MINGW", but MSYS2 subsystems like UCRT64 ship a
+    # Windows-native Node where os.type() returns "Windows_NT". In both cases
+    # MSYS2 sets the MSYSTEM env var (MINGW32, MINGW64, UCRT64, CLANG64, …).
+    # Either signal means GCC is the toolchain, so .S files compile directly
     # without the MSVC-specific preprocess_asm.cmd pre-processing step.
-    'target_platform%': '<!(node -p "require(\'os\').type().startsWith(\'MINGW\') ? \'mingw\' : (process.platform === \'win32\' ? \'msvc\' : \'\')")',
+    'target_platform%': '<!(node -p "require(\'os\').type().startsWith(\'MINGW\') || process.env.MSYSTEM ? \'mingw\' : (process.platform === \'win32\' ? \'msvc\' : \'\')")',
   },
   'target_defaults': {
     'default_configuration': 'Debug',

--- a/deps/libffi/libffi.gyp
+++ b/deps/libffi/libffi.gyp
@@ -9,6 +9,10 @@
 {
   'variables': {
     'target_arch%': 'ia32', # built for a 32-bit CPU by default
+    # Distinguish MSVC from MinGW/MSYS2 on Windows. MinGW reports an OS type
+    # starting with "MINGW" and uses GCC toolchain, so .S files compile directly
+    # without the MSVC-specific preprocess_asm.cmd pre-processing step.
+    'target_platform%': '<!(node -p "require(\'os\').type().startsWith(\'MINGW\') ? \'mingw\' : (process.platform === \'win32\' ? \'msvc\' : \'\')")',
   },
   'target_defaults': {
     'default_configuration': 'Debug',
@@ -50,9 +54,10 @@
     ],
   },
 
-  # Compile .S files on Windows
+  # MSVC-specific rules for preprocessing .preasm -> .asm -> .obj.
+  # MinGW uses GCC and compiles .S files directly without these rules.
   'conditions': [
-    ['OS=="win"', {
+    ['OS=="win" and target_platform=="msvc"', {
       'target_defaults': {
         'conditions': [
           ['target_arch=="ia32"', {
@@ -151,8 +156,11 @@
             ['OS=="linux" or OS=="mac"', {
               'sources': [ 'src/aarch64/sysv.S' ]
             }],
-            ['OS=="win"', {
+            ['OS=="win" and target_platform=="msvc"', {
               'sources': [ 'src/aarch64/win64_armasm.preasm' ]
+            }],
+            ['OS=="win" and target_platform=="mingw"', {
+              'sources': [ 'src/aarch64/sysv.S' ]
             }],
           ]
         }, { # ia32 or x64
@@ -160,29 +168,39 @@
             ['target_arch=="ia32"', {
               'sources': [ 'src/x86/ffi.c' ],
               'conditions': [
-                ['OS=="win"', {
+                ['OS=="win" and target_platform=="msvc"', {
                   'sources': [ 'src/x86/sysv_intel.preasm' ],
-                }, {
+                }],
+                ['OS=="win" and target_platform=="mingw"', {
+                  'sources': [ 'src/x86/sysv.S' ],
+                }],
+                ['OS!="win"', {
                   'sources': [ 'src/x86/sysv.S' ],
                 }],
               ],
             }],
             ['target_arch=="x64"', {
-              'sources': [
-                'src/x86/ffiw64.c',
-              ],
               'conditions': [
-                ['OS=="win"', {
+                ['OS=="win" and target_platform=="msvc"', {
                   'sources': [
+                    'src/x86/ffiw64.c',
                     'src/x86/win64_intel.preasm',
                   ],
-                }, {
+                  'msvs_disabled_warnings': [ 4267 ],
+                }],
+                ['OS=="win" and target_platform=="mingw"', {
+                  'sources': [
+                    'src/x86/ffiw64.c',
+                    'src/x86/win64.S',
+                  ],
+                }],
+                ['OS!="win"', {
                   'sources': [
                     'src/x86/ffi64.c',
                     'src/x86/unix64.S',
                     'src/x86/win64.S',
                   ],
-                }]
+                }],
               ],
             }],
             ['target_arch=="s390x"', {
@@ -190,11 +208,6 @@
                 'src/s390/ffi.c',
                 'src/s390/sysv.S',
               ],
-            }],
-            ['OS=="win"', {
-              # the libffi dlmalloc.c file has a bunch of implicit conversion
-              # warnings, and the main ffi.c file contains one, so silence them
-              'msvs_disabled_warnings': [ 4267 ],
             }],
           ]
         }],

--- a/deps/libffi/libffi.gyp
+++ b/deps/libffi/libffi.gyp
@@ -11,11 +11,12 @@
     'target_arch%': 'ia32', # built for a 32-bit CPU by default
     # Distinguish MSVC from MinGW/MSYS2 on Windows. MinGW-compiled Node reports
     # an OS type starting with "MINGW", but MSYS2 subsystems like UCRT64 ship a
-    # Windows-native Node where os.type() returns "Windows_NT". In both cases
-    # MSYS2 sets the MSYSTEM env var (MINGW32, MINGW64, UCRT64, CLANG64, …).
-    # Either signal means GCC is the toolchain, so .S files compile directly
-    # without the MSVC-specific preprocess_asm.cmd pre-processing step.
-    'target_platform%': '<!(node -p "require(\'os\').type().startsWith(\'MINGW\') || process.env.MSYSTEM ? \'mingw\' : (process.platform === \'win32\' ? \'msvc\' : \'\')")',
+    # Windows-native Node where os.type() returns "Windows_NT". Real MSYS2 sets
+    # MSYSTEM_PREFIX (e.g. /mingw64, /ucrt64); Git for Windows also sets MSYSTEM
+    # but does NOT set MSYSTEM_PREFIX, so we use that to avoid false positives
+    # in Git Bash environments. Either signal means GCC is the toolchain, so .S
+    # files compile directly without the MSVC preprocess_asm.cmd step.
+    'target_platform%': '<!(node -p "require(\'os\').type().startsWith(\'MINGW\') || process.env.MSYSTEM_PREFIX ? \'mingw\' : (process.platform === \'win32\' ? \'msvc\' : \'\')")',
   },
   'target_defaults': {
     'default_configuration': 'Debug',

--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "@napi-ffi/ffi-napi",
-  "version": "4.0.4",
+  "version": "4.0.5",
   "license": "MIT",
   "author": "Anna Henningsen <anna@addaleax.net>",
   "contributors": [
@@ -33,7 +33,7 @@
     "@napi-ffi/get-uv-event-loop-napi-h": "^1.0.6",
     "node-addon-api": "^8.2.1",
     "node-gyp-build": "^4.8.2",
-    "@napi-ffi/ref-napi": "^3.0.6",
+    "@napi-ffi/ref-napi": "^3.0.7",
     "@napi-ffi/ref-struct-di": "^1.1.1"
   },
   "devDependencies": {

--- a/test/library.js
+++ b/test/library.js
@@ -21,19 +21,22 @@ describe('Library', function () {
   });
 
   it('should accept `null` as a first argument', function () {
-    if (process.platform == 'win32') {
-      // On Windows, null refers to just the main executable (e.g. node.exe).
-      // Windows never searches for symbols across multiple DLL's.
-      // Note: Exporting symbols from an executable is unusual on Windows.
-      // Normally you only see exports from DLL's. It happens that node.exe
-      // does export symbols, so null as a first argument can be tested.
-      // This is an implementation detail of node, and could potentially
-      // change in the future (e.g. if node gets broken up into DLL's
-      // rather than being one large static linked executable).
+    if (process.platform == 'win32' && !process.env.MSYSTEM_PREFIX) {
+      // On Windows (MSVC builds), null refers to just the main executable
+      // (e.g. node.exe). Windows never searches for symbols across DLLs.
+      // node.exe happens to export uv_fs_open, so null can be tested here.
+      // This is an implementation detail of node that does not hold for
+      // MSYS2/MinGW builds where libuv may be a separate DLL or symbols
+      // are not re-exported from the node binary.
       const winFuncs = new Library(null, {
         'uv_fs_open': [ 'void', [ charPtr ] ]
       });
       assert(typeof winFuncs.uv_fs_open === 'function');
+    } else if (process.platform == 'win32') {
+      // MinGW/MSYS2: null is supported but uv_fs_open is not guaranteed
+      // to be exported from the node binary, skip the symbol check.
+      const l = new Library(null, {});
+      assert(typeof l === 'object');
     } else {
       // On POSIX, null refers to the global symbol table, and lets you use
       // symbols in the main executable and loaded shared libaries.

